File: de.raphaelmuesseler.financer.client/src/main/java/de/raphaelmuesseler/financer/client/format/I18N.java
package de.raphaelmuesseler.financer.client.format;

import de.raphaelmuesseler.financer.client.local.LocalStorage;
import de.raphaelmuesseler.financer.shared.model.user.User;

import java.text.MessageFormat;
import java.util.*;

/**
 * I18N utility class..
 *
 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
public final class I18N {
    /**
     * the current selected Locale.
     */
    private static LocalStorage localStorage;

    public static void setLocalStorage(LocalStorage localStorage) {
        I18N.localStorage = localStorage;
    }

    public enum Language {
        ENGLISH("English", Locale.ENGLISH),
        GERMAN("German", Locale.GERMAN);

        private final String name;
        private final Locale locale;

        Language(String name, Locale locale) {
            this.name = name;
            this.locale = locale;
        }

        public String getName() {
            return name;
        }

        public Locale getLocale() {
            return locale;
        }

        public static List<Language> getAll() {
            return Arrays.asList(values());
        }

        public static Language getLanguageByLocale(Locale locale) {
            for (Language language : values()) {
                if (language.getLocale().equals(locale)) {
                    return language;
                }
            }
            return null;
        }

        @Override
        public String toString() {
            return this.name;
        }
    }

    /**
     * getObject the supported Locales.
     *
     * @return List of Locale objects.
     */
    private static List<Locale> getSupportedLocales() {
        return new ArrayList<>(Arrays.asList(Locale.ENGLISH, Locale.GERMAN));
    }

    /**
     * getObject the default locale. This is the systems default if contained in the supported locales, english otherwise.
     *
     * @return default locale (en)
     */
    private static Locale getDefaultLocale() {
        Locale sysDefault = Locale.getDefault();
        return getSupportedLocales().contains(sysDefault) ? sysDefault : Locale.ENGLISH;
    }

    private static Locale getLocale() {
        if (I18N.localStorage.readObject("user") != null) {
            return ((User) I18N.localStorage.readObject("user")).getSettings().getLanguage();
        } else {
            return Locale.ENGLISH;
        }
    }

    /**
     * gets the string with the given key from the resource bundle for the current locale and uses it as first argument
     * to MessageFormat.format, passing in the optional args and returning the result.
     *
     * @param key  message key
     * @param args optional arguments for the message
     * @return localized formatted string
     */
    public static String get(final String key, final Object... args) {
        ResourceBundle bundle = ResourceBundle.getBundle("Financer", getLocale());
        return MessageFormat.format(bundle.getString(key), args);
    }
}

File: java/org.financer.client/src/main/kotlin/org/financer/client/format/I18N.kt
package org.financer.client.format

import org.financer.client.domain.model.user.User
import org.financer.client.local.LocalStorage
import org.financer.shared.domain.model.value.objects.SettingPair
import java.io.Serializable
import java.text.MessageFormat
import java.util.*

/**
 * I18N utility class..
 *
 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
object I18N {
    /**
     * the current selected Locale.
     */
    private var localStorage: LocalStorage? = null

    @kotlin.jvm.JvmStatic
    fun setLocalStorage(localStorage: LocalStorage?) {
        I18N.localStorage = localStorage
    }

    /**
     * getObject the supported Locales.
     *
     * @return List of Locale objects.
     */
    private val supportedLocales: List<Locale>
        get() = ArrayList(listOf(Locale.ENGLISH, Locale.GERMAN))

    /**
     * getObject the default locale. This is the systems default if contained in the supported locales, english
     * otherwise.
     *
     * @return default locale (en)
     */
    private val defaultLocale: Locale
        get() {
            val sysDefault = Locale.getDefault()
            return if (supportedLocales.contains(sysDefault)) sysDefault else Locale.ENGLISH
        }

    private val locale: Locale
        get() = if (localStorage!!.readObject<Serializable?>("user") != null) {
            (localStorage!!.readObject<Serializable>("user") as User).getValueOrDefault(SettingPair.Property.LANGUAGE)
        } else {
            SettingPair.Property.LANGUAGE.defaultValue as Locale
        }

    /**
     * gets the string with the given key from the resource bundle for the current locale and uses it as first argument
     * to MessageFormat.format, passing in the optional args and returning the result.
     *
     * @param key  message key
     * @param args optional arguments for the message
     * @return localized formatted string
     */
    @kotlin.jvm.JvmStatic
    operator fun get(key: String?, vararg args: Any?): String {
        val bundle = ResourceBundle.getBundle("Financer", locale)
        return MessageFormat.format(bundle.getString(key), *args)
    }

    enum class Language(val language: String, val locale: Locale) {
        ENGLISH("English", Locale.ENGLISH), GERMAN("German", Locale.GERMAN);

        override fun toString(): String {
            return language
        }

        companion object {
            @kotlin.jvm.JvmStatic
            val all: List<Language>
                get() = listOf(*values())

            @kotlin.jvm.JvmStatic
            fun getLanguageByLocale(locale: Locale): Language? {
                for (language in values()) {
                    if (language.locale == locale) {
                        return language
                    }
                }
                return null
            }
        }

    }
}